core/modules/replica/QservMgtServices.cc

```diff
@@ -0,0 +1,342 @@
+                manager->finish(request->id());
+            }
+        );
+
+    // Register the request (along with its callback) by its unique
+    // identifier in the local registry. Once it's complete it'll
+    // be automatically removed from the Registry.
+    _registry[request->id()] =
+        std::make_shared<QservMgtRequestWrapperImpl<GetReplicasQservMgtRequest>>(
+            request, onFinish);
+
+    // Initiate the request
+    request->start(service,
+                   jobId,
+                   requestExpirationIvalSec);
+
+    return request;
+}
+
+
+SetReplicasQservMgtRequest::Ptr QservMgtServices::setReplicas(
+                                        std::string const& worker,
+                                        QservReplicaCollection const& newReplicas,
+                                        bool force,
+                                        std::string const& jobId,
+                                        SetReplicasQservMgtRequest::CallbackType onFinish,
+                                        unsigned int requestExpirationIvalSec) {
+
+    util::Lock lock(_mtx, "QservMgtServices::setReplicas");
+
+    // Ensure we have the XROOTD/SSI service object before attempting any
+    // operations on requests
+    XrdSsiService* service = xrdSsiService();
+    if (not service) {
+        return SetReplicasQservMgtRequest::Ptr();
+    }
+
+    auto const manager = shared_from_this();
+
+    SetReplicasQservMgtRequest::Ptr const request =
+        SetReplicasQservMgtRequest::create(
+            _serviceProvider,
+            _io_service,
+            worker,
+            newReplicas,
+            force,
+            [manager] (QservMgtRequest::Ptr const& request) {
+                manager->finish(request->id());
+            }
+        );
+
+    // Register the request (along with its callback) by its unique
+    // identifier in the local registry. Once it's complete it'll
+    // be automatically removed from the Registry.
+    _registry[request->id()] =
+        std::make_shared<QservMgtRequestWrapperImpl<SetReplicasQservMgtRequest>>(
+            request, onFinish);
+
+    // Initiate the request
+    request->start(service,
+                   jobId,
+                   requestExpirationIvalSec);
+
+    return request;
+}
+
+void QservMgtServices::finish(std::string const& id) {
+
+    // IMPORTANT:
+    //
+    //   Make sure the notification is complete before removing
+    //   the request from the registry. This has two reasons:
+    //
+    //   - it will avoid a possibility of deadlocking in case if
+    //     the callback function to be notified will be doing
+    //     any API calls of the service manager.
+    //
+    //   - it will reduce the controller API dead-time due to a prolonged
+    //     execution time of of the callback function.
+
+    QservMgtRequestWrapper::Ptr request;
+    {
+        util::Lock lock(_mtx, "QservMgtServices::finish");
+        request = _registry[id];
```

Comment:
is there any case where this could be called when `id` is no longer (or never was) a key in the registry?

Reply:
In theory this should never be the case because "finish()" is called by a request which is done with a requested operation. However, I 100% agree with your concern, so I'm reinforcing the implementation to throw an exception if bo request is found in the registry for a particular identifier.
Thank you of the hint!